### ISSUE TITLE
Error found in mapping after migration from 11.0.1 to 12.0.1

### ISSUE THREAD

**Issue by marlonedu** [AutoMapper/AutoMapper#4361]
### Source/destination types
```csharp
public class ItemDeArbolDeActivosDeMedicion
{
    public long Id { get; }
    public string? Nombre { get; set; }
    public string? Abreviatura { get; set; }
    public byte[]? Imagen { get; set; }
    public int? IdPlataforma { get; set; }
    public int IdSecundario { get; set; }
    public IList<ItemDeArbolDeActivosDeMedicion> SubItems { get; set; } = new List<ItemDeArbolDeActivosDeMedicion>();

    public ItemDeArbolDeActivosDeMedicion? ItemPadre { get; private set; }

    public ItemDeArbolDeActivosDeMedicion(long id, TipoDeItemDeArbol? tipo = null)
    {
        this.Id = id;
        this.Tipo = tipo;
    }

    public TipoDeItemDeArbol? Tipo { get; }

    public void AgregarSubItems(params ItemDeArbolDeActivosDeMedicion[] subItems)
    {
        foreach (var subItem in subItems)
        {
            this.AgregarSubItemInterno(subItem);
        }
    }

    private void AgregarSubItemInterno(ItemDeArbolDeActivosDeMedicion subItem)
    {
        this.SubItems.Add(subItem);
        subItem.ItemPadre = this;
    }

    public override bool Equals(object obj)
    {
        if (obj is ItemDeArbolDeActivosDeMedicion otroItem)
        {
            return this.Equals(otroItem);
        }

        return false;
    }

    protected bool Equals(ItemDeArbolDeActivosDeMedicion other)
    {
        return this.Id == other.Id && this.Tipo == other.Tipo && this.IdPlataforma == other.IdPlataforma;
    }

    public override int GetHashCode()
    {
        unchecked
        {
            return (this.Id.GetHashCode() * 397) ^ this.Tipo.GetHashCode();
        }
    }

    public static bool operator ==(ItemDeArbolDeActivosDeMedicion? left, ItemDeArbolDeActivosDeMedicion? right)
    {
        return Equals(left, right);
    }

    public static bool operator !=(ItemDeArbolDeActivosDeMedicion? left, ItemDeArbolDeActivosDeMedicion? right)
    {
        return !Equals(left, right);
    }
}

public enum TipoDeItemDeArbol
{
    Cliente,
    Organizacion,
    Plataforma,
    Area,
    AreaDeAnalisis,
    SistemaDeMedicion,
    Variable,
    TipoDeParametroDeFluido,
    Tag,
    TagDeArea,
    VinculoEntreTagConSamplingPoint
}

public class ItemDeActivosDeMedicion_Vm
{
    public long Id { get; set; }
    public string? Nombre { get; set; }
    public string? Abreviatura { get; set; }
    public TipoDeItemDeArbol Tipo { get; set; }
    public byte[]? Imagen { get; set; }
    public int IdSecundario { get; set; }

    public ItemDeActivosDeMedicion_Vm? ItemPadre { get; set; }

    public IList<ItemDeActivosDeMedicion_Vm>? Items { get; set; }
    public int IdPlataforma { get; set; }
    public bool SeSelecciono { get; set; }
}
```

### Mapping configuration

```csharp
public class ItemDeActivosDeMedicion_Vm__Profile : Profile
{
    public ItemDeActivosDeMedicion_Vm__Profile()
    {
        // Dominio => Vista.
        this.CreateMap<ItemDeArbolDeActivosDeMedicion, ItemDeActivosDeMedicion_Vm>()
            .ForMember(vm => vm.Items, opt => opt.MapFrom(model => model.SubItems));
    }
}
```

### Expected behavior

The expected mapped objects should be equal in version 12.0.1 with 11.0.1, but some of the items in the recursion are not mapped properly.

### Actual behavior
The items in the mapped objects are mapped incorrectly for some of the items in the recursive structure.

### Steps to reproduce
I've uploaded a solution in github with the unit test written in xunit. There are two xunit projects: one with automapper version 12 and the other with automapper version 11. Each project has the same configuration, same types and the same data, the only difference is the installed automapper version. 

After running the test, you will see that the items for the type "Tag" are mapped incorrectly.

link to the code: https://github.com/marlonedu/automapper-tests

```csharp
// Your calls to Mapper.Map or ProjectTo here, with source/destination objects constructed
```

**Comment by lbargaoanu** [comment]
A repro would help. Make a [gist](https://gist.github.com/lbargaoanu/9c7233441c3a3413cc2b9b9ebb5964a9) that we can execute and see fail.
Simplify your code and reduce it to a repro with a few lines as requested.

**Comment by marlonedu** [comment]
There is no exception thrown, but there is an error in the objects mapped. Can't you check the attached project and see the failing unit test?

**Comment by lbargaoanu** [comment]
You can also try the [MyGet](https://docs.automapper.org/en/latest/The-MyGet-build.html) build.

**Comment by marlonedu** [comment]
The bug I found is fixed with the version 12.0.2-preview.0.4.
